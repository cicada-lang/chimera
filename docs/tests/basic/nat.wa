Nat "zero"

Nat { prev }
------------ {
  Nat prev
}

Zero "zero"

One { prev }
------------ {
  Zero prev
}

Two { prev }
------------ {
  One prev
}

Add ["zero", y, y]

Add [{ prev }, y, { prev: result_prev } ]
---------------------------------------- {
  Add [prev, y, result_prev]
}

query z {
  One x
  One y
  Add [x, y, z]
}

query (x, y, z) limit 1 {
  Add [x, y, z]
}

query (x, y, z) limit 3 {
  Add [x, y, z]
}

query (x, y) {
  Two z
  Add [x, y, z]
}

Mul ["zero", y, "zero"]

Mul [{ prev }, "zero", "zero"]

Mul [{ prev: prev_x }, { prev: prev_y } , result]
---------------------------- {
  Add [{ prev: prev_y }, z, result]
  Mul [prev_x, { prev: prev_y }, z]
}

query z {
  Two x
  Two y
  Mul [x, y, z]
}

query (x, y) {
  Two z
  Mul [x, y, z]
}
