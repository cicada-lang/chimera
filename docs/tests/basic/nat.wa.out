[ { "prev": { "prev": "zero" } } ]
[ ["zero", "?z", "?z"] ]
[ ["zero", "?z", "?z"], 
  [{ "prev": "zero" }, "?result_prev_17", { "prev": "?result_prev_17" }], 
  [{ "prev": { "prev": "zero" } }, "?result_prev_21", { "prev": { "prev": "?result_prev_21" } }] ]
[ ["zero", { "prev": { "prev": "zero" } }], 
  [{ "prev": "zero" }, { "prev": "zero" }], 
  [{ "prev": { "prev": "zero" } }, "zero"] ]
[ { "prev": { "prev": { "prev": { "prev": "zero" } } } } ]
[ [{ "prev": "zero" }, { "prev": { "prev": "zero" } }], 
  [{ "prev": { "prev": "zero" } }, { "prev": "zero" }] ]
